pkg/ruler/api.go

```diff
@@ -0,0 +1,260 @@
+	// Type of a recordingRule is always "recording".
+	Type string `json:"type"`
+}
+
+func respondError(logger log.Logger, w http.ResponseWriter, msg string, errType errorType) {
+	b, err := json.Marshal(&response{
+		Status:    "error",
+		ErrorType: errType,
+		Error:     msg,
+		Data:      nil,
+	})
+
+	if err != nil {
+		level.Error(logger).Log("msg", "error marshaling json response", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusInternalServerError)
+	if n, err := w.Write(b); err != nil {
+		level.Error(logger).Log("msg", "error writing response", "bytesWritten", n, "err", err)
+	}
+}
+
+func (r *Ruler) rules(w http.ResponseWriter, req *http.Request) {
+	logger := util.WithContext(req.Context(), util.Logger)
+	userID, ctx, err := user.ExtractOrgIDFromHTTPRequest(req)
+	if err != nil {
+		level.Error(logger).Log("msg", "error extracting org id from context", "err", err)
+		respondError(logger, w, "no valid org id found", errorInternal)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	rgs, err := r.GetRules(ctx, userID)
+
+	if err != nil {
+		respondError(logger, w, err.Error(), errorInternal)
+		return
+	}
+
+	groups := make([]*RuleGroup, 0, len(rgs))
+
+	for _, g := range rgs {
+		grp := RuleGroup{
+			Name:     g.Name,
+			File:     g.Namespace,
+			Interval: g.Interval.Seconds(),
+			Rules:    make([]rule, len(g.Rules)),
+		}
+
+		for i, rl := range g.Rules {
+			if g.Rules[i].Alert != "" {
+				alerts := make([]*Alert, 0, len(rl.Alerts))
+				for _, a := range rl.Alerts {
+					alerts = append(alerts, &Alert{
+						Labels:      client.FromLabelAdaptersToLabels(a.Labels),
+						Annotations: client.FromLabelAdaptersToLabels(a.Annotations),
+						State:       a.GetState(),
+						ActiveAt:    a.ActiveAt,
+						Value:       fmt.Sprintf("%f", a.Value),
+					})
+				}
+				grp.Rules[i] = alertingRule{
+					State:       rl.GetState(),
+					Name:        rl.GetAlert(),
+					Query:       rl.GetExpr(),
+					Duration:    rl.For.Seconds(),
+					Labels:      client.FromLabelAdaptersToLabels(rl.Labels),
+					Annotations: client.FromLabelAdaptersToLabels(rl.Annotations),
+					Alerts:      alerts,
+					Health:      rl.GetHealth(),
+					LastError:   rl.GetLastError(),
+					Type:        "alerting",
+				}
+			} else {
+				grp.Rules[i] = recordingRule{
+					Name:      rl.GetRecord(),
+					Query:     rl.GetExpr(),
+					Labels:    client.FromLabelAdaptersToLabels(rl.Labels),
+					Health:    rl.GetHealth(),
+					LastError: rl.GetLastError(),
+					Type:      "recording",
+				}
+			}
+		}
+		groups = append(groups, &grp)
+	}
+
+	b, err := json.Marshal(&response{
+		Status: "success",
+		Data:   &RuleDiscovery{RuleGroups: groups},
+	})
+	if err != nil {
+		level.Error(logger).Log("msg", "error marshaling json response", "err", err)
+		respondError(logger, w, "unable to marshal the requested data", errorBadData)
```

Comment:
Why is this a `errorBadData`? If we're able to marshal the response looks more a bug on our side. Moreover, the bad data error is mapped to a `400` response status code, while we always return 5xx, but I think it's correct 5xx here since there's nothing the user can do on it.

_Same comment applies to `alerts()`_

Reply:
👍